SecurityManager/Core.UnitTests/Persistence/SecurityManagerSqlStorageObjectFactoryTest.cs

```diff
@@ -80,22 +80,24 @@ public void CreateReadOnlyStorageProvider ()
     {
       var result = _securityManagerSqlStorageObjectFactory.CreateReadOnlyStorageProvider(_rdbmsProviderDefinition, _persistenceExtensionStub.Object);
 
-      Assert.That(result, Is.TypeOf(typeof(SecurityManagerRdbmsProvider)));
-      Assert.That(result.As<SecurityManagerRdbmsProvider>().PersistenceExtension, Is.SameAs(_persistenceExtensionStub.Object));
-      Assert.That(result.As<SecurityManagerRdbmsProvider>().StorageProviderDefinition, Is.SameAs(_rdbmsProviderDefinition));
-      Assert.That(result.As<SecurityManagerRdbmsProvider>().ConnectionString, Is.EqualTo(_rdbmsProviderDefinition.ReadOnlyConnectionString));
+      Assert.That(result, Is.TypeOf(typeof(ReadOnlyStorageProviderDecorator)));
+      var innerStorageProvider = result.As<ReadOnlyStorageProviderDecorator>().InnerStorageProvider;
+      Assert.That(innerStorageProvider.As<SecurityManagerRdbmsProvider>().PersistenceExtension, Is.SameAs(_persistenceExtensionStub.Object));
+      Assert.That(innerStorageProvider.As<SecurityManagerRdbmsProvider>().StorageProviderDefinition, Is.SameAs(_rdbmsProviderDefinition));
+      Assert.That(innerStorageProvider.As<SecurityManagerRdbmsProvider>().ConnectionString, Is.EqualTo(_rdbmsProviderDefinition.ReadOnlyConnectionString));
     }
 
     [Test]
     public void CreateReadOnlyStorageProviderWithMixin ()
     {
+      //I have no clue how to fix this test. Any input?
       using (
           MixinConfiguration.BuildFromActive().ForClass(typeof(RdbmsProvider)).Clear().AddMixins(typeof(SecurityManagerRdbmsProviderTestMixin)).
               EnterScope())
       {
         var result = _securityManagerSqlStorageObjectFactory.CreateReadOnlyStorageProvider(_rdbmsProviderDefinition, _persistenceExtensionStub.Object);
 
-        Assert.That(Mixin.Get<SecurityManagerRdbmsProviderTestMixin>(result), Is.Not.Null);
+        Assert.That(Mixin.Get<RdbmsProvider>(result), Is.Not.Null);
```

Comment:
The `result` is the `ReadOnlyStorageProviderDecorator`; you need the decorated `SecurityManagerRdbmsProvider` for the call to `Mixin.Get<SecurityManagerRdbmsProviderTestMixin>()`.
This is because `RdbmsProvider` - and therefore its derived class `SecurityManagerRdbmsProvider` - is the mixin target, which is decorated within `SqlStorageObjectFactory.GetReadOnlyStorageProvider()`.

Reply:
For the updated: I'm not sure this is how we actually want the API to look, but I'm probably missing something.

Reply:
That is fine. We don't usually want to get the `Mixin` objects, we only do that here to test that the instance is indeed mixed, i.e. created using `ObjectFactory` (instead of `new`). Since the mixed instance is now decorated, the test must cast to the decorator to get to the mixed object.
For developers using the `IReadOnlyStorageProvider`, these details are hidden behind the interface, and everything works as intended.

Reply:
Ok, thanks for the explanation! :)